### PR TITLE
feat(eslint-plugin): rework configs: recommended, strict, stylistic; -type-checked

### LINKED ISSUE #5204: Configs: Have recommended/strict configs include lesser configs
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Description

Right now, no config in the following list does not include any other config in that list:

* `plugin:@typescript-eslint/recommended`
* `plugin:@typescript-eslint/recommended-requiring-type-checking`
* `plugin:@typescript-eslint/strict` 

That means if you want, say, the strictest, you must enable all three (https://typescript-eslint.io/docs/linting/configs):

```js
{
  "extends": [
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "plugin:@typescript-eslint/strict"
  ]
}
```

Proposal: how about we have each config in that list also include any previous config in the list? That way if you want, say, the strictest, you would only need to enable one:

```js
{
  "extends": [
    "plugin:@typescript-eslint/strict"
  ]
}
```

### Impacted Configurations

* `plugin:@typescript-eslint/recommended`
* `plugin:@typescript-eslint/recommended-requiring-type-checking`
* `plugin:@typescript-eslint/strict` 

### Additional Info

This would be a breaking change.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
BREAKING CHANGE:
changes the recommended configs

---

## PR Checklist

-   [x] Addresses an existing open issue: fixes #5204
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Has each config after `recommended` also include all rules from prior configs. Simplifies what users will have to configure by letting them only specify the strictest desired config.

This is being used to generate the table in #6014.

### CODE REVIEW

File: packages/eslint-plugin/tools/generate-configs.ts [hunk 1]
```diff
@@ -9,7 +9,10 @@ function addAutoGeneratedComment(code: string): string {
...
     '// THIS CODE WAS AUTOMATICALLY GENERATED',
     '// DO NOT EDIT THIS CODE BY HAND',
-    '// YOU CAN REGENERATE IT USING yarn generate:configs',
+    '// SEE https://typescript-eslint.io/docs/linting/configs',
+    '//',
+    '// For developers working in the typescript-eslint monorepo:',
+    '// You can regenerate it using `yarn generate:configs`',
     '',
     code,
   ].join('\n');
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tools/generate-configs.ts:18]: By the way, I get this lovely error from `chalk` when I try to `yarn generate:configs`:

```plaintext
$ ../../node_modules/.bin/ts-node --files --transpile-only tools/generate-configs.ts
/home/josh/repos/typescript-eslint/packages/eslint-plugin/tools/generate-configs.ts:6
import rules from '../src/rules';
                                ^
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/josh/repos/typescript-eslint/packages/eslint-plugin/node_modules/chalk/source/index.js from /home/josh/repos/typescript-eslint/packages/eslint-plugin/tools/generate-configs.ts not supported.
Instead change the require of index.js in /home/josh/repos/typescript-eslint/packages/eslint-plugin/tools/generate-configs.ts to a dynamic import() which is available in all CommonJS modules.
    at Object.<anonymous> (/home/josh/repos/typescript-eslint/packages/eslint-plugin/tools/generate-configs.ts:6:33)
    at Module.m._compile (/home/josh/repos/typescript-eslint/node_modules/ts-node/dist/index.js:791:29)
    at Object.require.extensions.<computed> [as .ts] (/home/josh/repos/typescript-eslint/node_modules/ts-node/dist/index.js:793:16)
    at phase4 (/home/josh/repos/typescript-eslint/node_modules/ts-node/dist/bin.js:407:16)
    at bootstrap (/home/josh/repos/typescript-eslint/node_modules/ts-node/dist/bin.js:49:12)
    at main (/home/josh/repos/typescript-eslint/node_modules/ts-node/dist/bin.js:32:12)
    at Object.<anonymous> (/home/josh/repos/typescript-eslint/node_modules/ts-node/dist/bin.js:519:5) {
  code: 'ERR_REQUIRE_ESM'
}
```

I resolved it locally by commenting out all the logging. I'm down to just remove the nice pretty printing for now.

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tools/generate-configs.ts:19]: #5925

**Comment by bradzacher** [comment]
converting to draft pending resolution of https://github.com/typescript-eslint/typescript-eslint/discussions/6019

File: package.json [hunk 2]
```diff
@@ -30,15 +30,15 @@
...
+    "generate-contributors": "node --loader ts-node/esm ./tools/generate-contributors.ts",
+    "generate-sponsors": "node --loader ts-node/esm ./tools/generate-sponsors.ts",
+    "generate-website-dts": "node --loader ts-node/esm ./tools/generate-website-dts.ts",
     "generate-lib": "nx generate-lib @typescript-eslint/scope-manager",
     "lint-fix": "eslint . --fix",
     "lint-markdown-fix": "yarn lint-markdown --fix",
     "lint-markdown": "markdownlint \"**/*.md\" --config=.markdownlint.json --ignore-path=.markdownlintignore",
     "lint": "nx run-many --target=lint --all --parallel",
-    "postinstall": "yarn ts-node --transpile-only ./tools/postinstall.ts",
+    "postinstall": "node --loader ts-node/esm ./tools/postinstall.ts",
```
Comment (JoshuaKGoldberg) [inline package.json:41]: 🤔 both with `yarn ts-node --transpile-only ...` and `node --loader ts-node/esm ...`, [CI fails](https://github.com/typescript-eslint/typescript-eslint/actions/runs/3653072424/jobs/6172135352) with:

```
Error: Cannot find module '../dist-raw/node-cjs-loader-utils'
Require stack:
- /home/runner/work/typescript-eslint/typescript-eslint/node_modules/ts-node/dist/index.js
- /home/runner/work/typescript-eslint/typescript-eslint/node_modules/ts-node/dist/esm.js
- /home/runner/work/typescript-eslint/typescript-eslint/node_modules/ts-node/esm.mjs
```

It doesn't repro locally for me.

File: packages/website/src/theme/MDXComponents/RuleAttributes.tsx [hunk 3]
```diff
@@ -6,12 +6,18 @@ import type { FeatureProps } from './Feature';
...
-const getRecommendation = (docs: RuleMetaDataDocs): [string, string] => {
-  return docs.recommended === 'strict'
-    ? ['🔒', 'strict']
-    : docs.requiresTypeChecking
-    ? ['🧠', 'recommended-requiring-type-checking']
-    : ['✅', 'recommended'];
+const recommendations = {
+  recommended: ['✅', 'recommended'],
+  strict: ['🔒', 'strict'],
+  stylistic: ['🎨', 'stylistic'],
```
Comment (JoshuaKGoldberg) [inline packages/website/src/theme/MDXComponents/RuleAttributes.tsx:12]: Per https://github.com/bmish/eslint-doc-generator/pull/136 

cc @bmish - thought you'd enjoy this 😄

**Comment by JoshuaKGoldberg** [comment]
😩 these integration tests and flaky builds are going to be the death of me...

File: .github/actions/prepare-install/action.yml [hunk 4]
```diff
@@ -46,8 +46,10 @@ runs:
       with:
         path: |
           ${{ env.yarn_cache_dir }}
-        key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+        key: ${{ runner.os }}-yarn-${{ hashFiles('**/package.json') }}-${{ hashFiles('**/yarn.lock') }}
         restore-keys: |
+          ${{ runner.os }}-yarn-${{ hashFiles('**/package.json') }}-${{ hashFiles('**/yarn.lock') }}
+          ${{ runner.os }}-yarn-${{ hashFiles('**/package.json') }}
```
Comment (JoshuaKGoldberg) [inline .github/actions/prepare-install/action.yml:52]: These were from me testing around cache fixes. They didn't work. #6246

**Review by bradzacher — CHANGES_REQUESTED** [review]

File: packages/utils/src/ts-eslint/Rule.ts [hunk 5]
```diff
@@ -17,7 +17,7 @@ interface RuleMetaDataDocs {
    * Used by the build tools to generate the recommended and strict configs.
    * Set to false to not include it as a recommendation
    */
-  recommended: 'error' | 'strict' | 'warn' | false;
+  recommended?: RuleRecommendation;
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/utils/src/ts-eslint/Rule.ts:20]: Agree. I think making it `?:` optional will help, so folks don't _have_ to apply it.

I thought there was an issue filed on the repo already but I can't find it...

File: packages/website/src/components/ast/serializer/serializerTS.ts [hunk 6]
```diff
@@ -47,8 +47,8 @@ function expandFlags(
   flags: number,
 ): string {
   return Object.entries(allFlags[1])
-    .filter(([f, _]) => (Number(f) & flags) !== 0)
-    .map(([_, name]) => `${allFlags[0]}.${name}`)
+    .filter(([f]) => (Number(f) & flags) !== 0)
+    .map(([name]) => `${allFlags[0]}.${name}`)
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/website/src/components/ast/serializer/serializerTS.ts:51]: I ... don't remember why this was changed, if it was intentional. 🤔

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/website/src/components/ast/serializer/serializerTS.ts:51]: Oh right I'd been fiddling with removing the `_`s. No need.

File: packages/eslint-plugin/src/index.ts [hunk 7]
```diff
@@ -3,17 +3,25 @@ import base from './configs/base';
...
 export = {
-  rules,
   configs: {
     all,
     base,
-    recommended,
     'eslint-recommended': eslintRecommended,
+    recommended,
     'recommended-requiring-type-checking': recommendedRequiringTypeChecking,
+    'recommended-type-checked': recommendedTypeChecked,
```
Comment (bradzacher) [inline packages/eslint-plugin/src/index.ts:20]: i was going to say you could save that one file and do this instead:
```suggestion
    'recommended-requiring-type-checking': recommendedTypeChecked,
    'recommended-type-checked': recommendedTypeChecked,
```

File: packages/utils/src/ts-eslint/Rule.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bradzacher) [reply packages/utils/src/ts-eslint/Rule.ts:20]: Do we want to make this a generic and just make it default to the old value?
In our `createRule` function we can overwrite it with our own type so we only do this locally.

**Review by bradzacher — APPROVED** [review]
futureisnow_oldman

File: packages/eslint-plugin/src/index.ts (hunk 7 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/index.ts:20]: Oh ha yeah true. And on top of this & v6 overall already being breaking changes, with #6458 we don't have to worry about people reaching into `dist/`. Great.

File: packages/utils/src/ts-eslint/Rule.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/utils/src/ts-eslint/Rule.ts:20]: I like it. Filed #6467.

**Pull request merged**